internal/xpkg/v2/build.go

```diff
@@ -0,0 +1,322 @@
+	"github.com/crossplane/crossplane-runtime/pkg/parser"
+
+	pkgmetav1 "github.com/crossplane/crossplane/apis/pkg/meta/v1"
+	"github.com/crossplane/crossplane/apis/pkg/meta/v1beta1"
+	xpkgv1 "github.com/crossplane/crossplane/internal/xpkg"
+	"github.com/crossplane/crossplane/internal/xpkg/v2/parser/examples"
+	"github.com/crossplane/crossplane/internal/xpkg/v2/scheme"
+)
+
+const (
+	errParserPackage     = "failed to parse package"
+	errParserExample     = "failed to parse examples"
+	errLintPackage       = "failed to lint package"
+	errInitBackend       = "failed to initialize package parsing backend"
+	errTarFromStream     = "failed to build tarball from stream"
+	errLayerFromTar      = "failed to convert tarball to image layer"
+	errDigestInvalid     = "failed to get digest from image layer"
+	errBuildImage        = "failed to build image from layers"
+	errConfigFile        = "failed to get config file from image"
+	errMutateConfig      = "failed to mutate config for image"
+	errBuildObjectScheme = "failed to build scheme for package encoder"
+	errParseAuth         = "an auth extension was supplied but could not be parsed"
+	errAuthNotAnnotated  = "an auth extension was supplied but but the " + providerConfigKind + " object could not be found"
+	errNotExactlyOneMeta = "package must contain exactly one meta object"
+	authMetaAnno         = "auth.upbound.io/group"
+	authObjectAnno       = "auth.upbound.io/config"
+	providerConfigKind   = "ProviderConfig"
+)
+
+// annotatedTeeReadCloser is a copy of io.TeeReader that implements
+// parser.AnnotatedReadCloser. It returns a Reader that writes to w what it
+// reads from r. All reads from r performed through it are matched with
+// corresponding writes to w. There is no internal buffering - the write must
+// complete before the read completes. Any error encountered while writing is
+// reported as a read error. If the underlying reader is a
+// parser.AnnotatedReadCloser the tee reader will invoke its Annotate function.
+// Otherwise it will return nil. Closing is always a no-op.
+func annotatedTeeReadCloser(r io.Reader, w io.Writer) *teeReader {
+	return &teeReader{r, w}
+}
+
+type teeReader struct {
+	r io.Reader
+	w io.Writer
+}
+
+func (t *teeReader) Read(p []byte) (n int, err error) {
+	n, err = t.r.Read(p)
+	if n > 0 {
+		if n, err := t.w.Write(p[:n]); err != nil {
+			return n, err
+		}
+	}
+	return
+}
+
+func (t *teeReader) Close() error {
+	return nil
+}
+
+func (t *teeReader) Annotate() any {
+	anno, ok := t.r.(parser.AnnotatedReadCloser)
+	if !ok {
+		return nil
+	}
+	return anno.Annotate()
+}
+
+// Builder defines an xpkg Builder.
+type Builder struct {
+	pb parser.Backend
+	eb parser.Backend
+	ab parser.Backend
+
+	pp parser.Parser
+	ep *examples.Parser
+}
+
+// New returns a new Builder.
+func New(pkg, ab, ex parser.Backend, pp parser.Parser, ep *examples.Parser) *Builder {
+	return &Builder{
+		pb: pkg,
+		ab: ab,
+		eb: ex,
+		pp: pp,
+		ep: ep,
+	}
+}
```

Comment:
some more documentation on the meaning of these would be nice, but we can leave that to a follow up PR 😅

Reply:
Made them a bit more self-explanatory